homeassistant/components/homekit/config_flow.py

```diff
@@ -89,51 +111,90 @@ class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+                    vol.Required(
+                        CONF_INCLUDE_DOMAINS, default=default_domains
+                    ): cv.multi_select(SUPPORTED_DOMAINS),
+                }
+            ),
+        )
+
     async def async_step_pairing(self, user_input=None):
         """Pairing instructions."""
         if user_input is not None:
-            return self.async_create_entry(
-                title=self.entry_title, data=self.homekit_data
-            )
+            return self.async_create_entry(title=self.entry_title, data=self.hk_data)
+
+        self.hk_data[CONF_PORT] = await self._async_available_port()
+        self.hk_data[CONF_NAME] = self._async_available_name(
+            self.hk_data[CONF_HOMEKIT_MODE]
+        )
+        self.entry_title = f"{self.hk_data[CONF_NAME]}:{self.hk_data[CONF_PORT]}"
         return self.async_show_form(
             step_id="pairing",
-            description_placeholders={CONF_NAME: self.homekit_data[CONF_NAME]},
+            description_placeholders={CONF_NAME: self.hk_data[CONF_NAME]},
         )
 
     async def async_step_user(self, user_input=None):
         """Handle the initial step."""
         errors = {}
-        if user_input is not None:
-            port = await self._async_available_port()
-            name = self._async_available_name()
-            title = f"{name}:{port}"
-            self.homekit_data = user_input.copy()
-            self.homekit_data[CONF_NAME] = name
-            self.homekit_data[CONF_PORT] = port
-            self.homekit_data[CONF_FILTER] = {
-                CONF_INCLUDE_DOMAINS: user_input[CONF_INCLUDE_DOMAINS],
-                CONF_INCLUDE_ENTITIES: [],
-                CONF_EXCLUDE_DOMAINS: [],
-                CONF_EXCLUDE_ENTITIES: [],
-            }
-            del self.homekit_data[CONF_INCLUDE_DOMAINS]
-            self.entry_title = title
-            return await self.async_step_pairing()
 
-        default_domains = [] if self._async_current_names() else DEFAULT_DOMAINS
-        setup_schema = vol.Schema(
-            {
-                vol.Required(
-                    CONF_INCLUDE_DOMAINS, default=default_domains
-                ): cv.multi_select(SUPPORTED_DOMAINS),
+        if user_input is not None:
+            self.hk_data = {
+                CONF_HOMEKIT_MODE: user_input[CONF_HOMEKIT_MODE],
             }
-        )
+            if user_input[CONF_HOMEKIT_MODE] == HOMEKIT_MODE_ACCESSORY:
+                return await self.async_step_accessory_mode()
+            return await self.async_step_bridge_mode()
 
+        homekit_mode = self.hk_data.get(CONF_HOMEKIT_MODE, DEFAULT_HOMEKIT_MODE)
         return self.async_show_form(
-            step_id="user", data_schema=setup_schema, errors=errors
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_HOMEKIT_MODE, default=homekit_mode): vol.In(
+                        HOMEKIT_MODES
+                    )
```

Comment:
Is this important enough to always show it to all users. Why don't we pick the preferred mode based on the domain?

Reply:
I thought about it but I couldn't come up with a workflow that guided the user to a good result without asking the mode first.  Right now the UX gets them in a state where they end up with bridged cameras because its the no thought click though action.  This puts them in a state where homekit is unstable since everything blocks while the camera takes a snapshot.  I wanted to guide them away from ending up in that state.

With accessory mode, you only get a single entity so it seemed like I needed to have them pick what mode first so we can get to selecting a single one vs selecting a whole domain.

Accessory Path: [accessory] -> [pick entity] -> Done
Bridge Path: [bridge] -> [select domains] -> Done

If we go the route of picking the domain first the accessory path get longer, which seemed painful when you have to setup a few cameras and it wasn't very discoverable because you would have to know to select one domain only on the first screen.

Accessory Path: [pick domain] -> [pick mode] -> [pick entity] -> Done
Bridge Path: [pick domain(s)] -> [pick mode] (skipped of no camera or media_players selected) -> Done

Reply:
What if we just assume bridge, and if they pick any camera, we create those automatically as individual config entries? We can use an approach similar to how we do config flow import to create the extra entries.

Reply:
I'm just trying to look at solutions where we can avoid putting the burden on the user of knowing how Homekit works.

Reply:
I like that idea but we can't hide too much of the detail from the user since they still have to pair each one. If you select 10 cameras and a few switches you end up with 11 qr codes.  

I might be a bit biased about them wanting them to know a little more about how it works since I end up frequently explaining the implementation details frequently in issues and on discord when users have availablity issues with homekit. I've got these for quick responses:

`Camera snapshots block the homekit connection from updating other devices. Please create a separate homekit instance in accessory mode for each camera. This can be done by going to Configuration >> Integrations in the UI, click the button with + sign and from the list of integrations select HomeKit. Select `camera`, choose include mode, choose the one you want, and complete the steps. Do not pair the bridge. Go back into Options and change the bridge to accessory mode, and complete the steps. There is currently a bug that will not setup the filter correctly, so you will need to go back into options, and complete the steps again. Now pair using the QR code.`

`TV Media players will display the wrong icon and randomly loose inputs if they are bridged due to homekit's implementation. This can be done by going to Configuration >> Integrations in the UI, click the button with + sign and from the list of integrations select HomeKit. Select `media_player`, choose include mode, choose the one you want, and complete the steps. Do not pair the bridge. Go back into Options and change the bridge to accessory mode, and complete the steps. There is currently a bug that will not setup the filter correctly, so you will need to go back into options, and complete the steps again. Now pair using the QR code.`

Reply:
What if we show the QR codes in the config flow with extra description why the user has to pair more times?

 - Select devices
 - Scan this QR code to pair Homekit
 - Thanks for pairing! It looks like you have selected to expose camera `camera.backyard`. You will have to pair that separately due to a limitation in Homekit. <Show QR code below this text> 
 - Repeat previous step until all devices done.

Reply:
I *think* that is doable, but its quite a refactor since we still need to support showing the existing notification when the bridge/device is unpaired so they can pair it again.  How about doing that as a next step?

Reply:
When testing I realized we don't show the name they will see in homekit in the pairing screen so I cleaned that up here https://github.com/home-assistant/core/pull/45665

Reply:
I'm trying to figure out how to generate multiple config entries from a single config flow. Since at the end the data is returned resulting in an entry, I haven't figured out a good way to make multiple.  Is there some prior art on this?

Reply:
You can do something similar to the "import" step we do to import YAML config. Just name it differently and create those for the camera imports.

Reply:
I should be able to get that done before next beta as long as the XSS protection doesn't cause any roadblocks